### ISSUE TITLE
Support multiple key-value pairs for arrays in HTTP API

### ISSUE THREAD

**Issue by michaelsproul** [sigp/lighthouse#2739]
## Description

In the HTTP API there are several places where we use comma-separated lists in query parameters e.g. for the `beacon/states/{state_id}/validators` API:

```bash
curl "http://localhost:5052/eth/v1/beacon/states/head/validators?id=0,1"
```

This corresponds to `style: form`, `explode: false` from the OpenAPI specification: https://swagger.io/docs/specification/serialization/#query

However, the default serialisation format is defined to be `style: form`, `explode: true`, which looks like this:

```bash
curl "http://localhost:5052/eth/v1/beacon/states/head/validators?id=0&id=1"
```

Currently this 2nd query will fail with a deserialisation error:

```json
{
  "code": 400,
  "message": "BAD_REQUEST: invalid query: Invalid query string",
  "stacktraces": []
}
```

## Steps to fix

I think we should adapt a hybrid style that avoids breaking backwards compatibility. We should collect all values of the key, and then comma-split each value and flatten the results into a single list. E.g.

```
id=1&id=2,3&id=4,5
```

parses as:

```
ids = [1, 2, 3, 4, 5]
```

and queries like:

```
id=1,2,3,4,5
```

parse the same as they do currently.

The (de)serialisation is implemented in the `QueryVec` type here:

https://github.com/sigp/lighthouse/blob/fff01b24ddedcd54486e374460855ca20d3dd232/common/eth2/src/types.rs#L523-L525

If we're lucky we might be able to get it to work by changing its `TryFrom` impl to `TryFrom<Vec<String>>` :crossed_fingers: 

## Version

Lighthouse v2.0.1

## Additional Info

Thanks to @guybrush for raising this (https://discord.com/channels/605577013327167508/605577013331361793/900420789365510144)

**Comment by mooori** [comment]
## Tl;dr
I've been looking into this and I'm stuck on getting a sequence of _all_ values of a repeated key to `TryFrom`: for query `id=1&id=2` I couldn't get `["1", "2"]` passed to `TryFrom`.

If this is possible, it would be much appreciated if someone could point out how to do it.

[Here](https://github.com/mooori/multikey-api) is a small project which isolates the issue of `QueryVec` deserialization.

## Findings / What I tried
Maybe this might help to find a solution I couldn't see.

### Find out more about error `Invalid query string`
`warp::filters::query` [uses](https://docs.rs/warp/0.3.2/src/warp/filters/query.rs.html#65-79) `serde_urlencoded`, which has chosen _not_ to support duplicate query parameters ([issue](https://github.com/nox/serde_urlencoded/issues/6), [issue](https://github.com/nox/serde_urlencoded/issues/75)).

Enabling tracing on `warp` requests shows that deserialization fails with `duplicate field` error. According to [this issue](https://github.com/serde-rs/serde/issues/690) that's a dead end.

### Using `TryFrom<Vec<String>>`
...via

```rust
#[serde(try_from = "Vec<String>", bound = "T: FromStr")]
```
results in error `Invalid query string`. Tracing the request in `warp` shows that `TryFrom<Vec<String>>` receives only a `String` (the first value of the repeated query param), not `Vec<String>` (all values) ([comment](https://github.com/nox/serde_urlencoded/issues/6#issuecomment-456732400)).

### Implement `Deserialize` for `QueryVec`
This is based on this [stackoverflow post](https://stackoverflow.com/a/64406395). Assume trying to deserialize `q=1&q=2` for

```rust
struct S {
    #[serde(flatten)]
    q: QueryVec<u64>
}
```
Deserialization fails with `Error("can only flatten structs and maps")`.

Probably due to `QueryVec<T>` having a generic parameter, [this](https://github.com/serde-rs/serde/pull/1771) `serde` issue or a combination of both.

It appears to be the wrong approach and I couldn't make it work. There might be a way I didn't see, though.

### `serde_with`
There is an [example](https://docs.rs/serde_with/1.11.0/serde_with/rust/tuple_list_as_map/index.html#examples) showing how duplicate keys can be deserialized with `tuple_list_as_map`. However, this fails in conjunction with `warp::query` as input must be a map, not just the string value of the key's first occurrence (see section `TryFrom<Vec<String>>`).

## Another possible approach: `serde_qs`
[`serde_qs`](https://docs.rs/serde_qs/0.8.5/serde_qs/index.html#supported-types) supports deserializing repeated key-value pairs as sequences. However, it requires such keys in the query string to be annotated with indices, like `id[0]=1&id[1]=2`. Hence I assume `serde_qs` can't help to close this issue.

**Comment by macladson** [comment]
Thank you for the thorough breakdown of your work @mooori. Having your notes on hand was extremely helpful while attempting to solve this myself.

As you noted in your comment, using `warp`'s default implementation which uses `serde_urlencoded`, was a dead-end as they (rightly so) are concered about the performance overhead of supporting something like this.

I went down a bit of a rabbit hole with trying to extend `serde_qs` to support the multi-key queries without needing the square brackets, however it seems as though the architecture of the library is ill-suited for something like that (as the [maintainers themselves also noted](https://github.com/samscott89/serde_qs/issues/16#issuecomment-450464628)).

Then I found this very helpful comment: [https://github.com/seanmonstar/warp/issues/733#issuecomment-722359432](https://github.com/seanmonstar/warp/issues/733#issuecomment-722359432)

This means we can parse out each key and value in the query string into a `Vec<(String, String)>`.  From there, we can deserialize into the necessary types.

To avoid needing to have a separate deserializing function for each type, @michaelsproul wrote a general deserializer for these types of query strings [here](https://github.com/sigp/serde_array_query).

And I'm using that library in my PR [here](https://github.com/sigp/lighthouse/pull/2908).

It ended up being more involved than we were hoping but thanks again @mooori for the work you did.

**Comment by mooori** [comment]
Great to hear it was helpful! I was indeed curious how this was solved, so thanks a lot for the breakdown of your solution @macladson.

**Comment by michaelsproul** [comment]
Closed by #2908
